googleapis/src/main/java/io/grpc/googleapis/GoogleCloudToProdNameResolver.java

```diff
@@ -162,16 +162,25 @@ public void run() {
         try {
           zone = queryZoneMetadata(METADATA_URL_ZONE);
           supportIpv6 = queryIpv6SupportMetadata(METADATA_URL_SUPPORT_IPV6);
-          rawBootstrap = generateBootstrap(zone, supportIpv6);
+
+          // User provided bootstrap configs are only supported with federation. If federation is
+          // not enabled or there is no user provided config, we set a custom bootstrap override.
+          // Otherwise, we don't set the override, which will allow a user provided bootstrap config
+          // to take effect.
+          if (!enableFederation || !xdsBootstrapProvided) {
```

Comment:
we can make this conditional also encompass the zone and IPv6 metadata queries, to avoid unnecessary I/O in the case that we're using a bootstrap

E.g.:

```
if (!enableFederation || !xdsBootstrapProvided) {
  zone = queryZoneMetadata(METADATA_URL_ZONE);
  supportIpv6 = queryIpv6SupportMetadata(METADATA_URL_SUPPORT_IPV6);
  rawBootstrap = generateBootstrap(zone, supportIpv6);
}
```

Reply:
Makes sense, I was too laser focused on the generation part. I moved the zone and ipv6 queries into the condition and did some refactoring as well to improve readability.